src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+  let i = h land mask in
+  assoc r key @@ Atomic_array.unsafe_fenceless_get r.buckets i
+(* *)
+
+let find_exn t key =
+  match find_node t key with
+  | Nil -> raise_notrace Not_found
+  | Cons r -> r.value
+
+let mem t key = find_node t key != Nil
+
+(* *)
+
+let rec try_add t key value backoff =
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  match Atomic_array.unsafe_fenceless_get r.buckets i with
+  | B (Nil_with_size nil_r as before) ->
+      if nil_r.size_modifier != Size.used_once then
+        Size.update_once r.size nil_r.size_modifier;
+      let (Cons_with_size cons_r as after : ('a, 'b, [ `Cons_with_size ]) tdt) =
+        Cons_with_size
+          {
+            size_modifier = Size.new_once r.size Size.incr;
+            key;
+            value;
+            rest = Nil;
+          }
+      in
+      if Atomic_array.unsafe_compare_and_set r.buckets i (B before) (B after)
+      then (
+        Size.update_once r.size cons_r.size_modifier;
+        cons_r.size_modifier <- Size.used_once;
+        adjust_size t r mask true)
+      else try_add t key value (Backoff.once backoff)
+  | B (Cons_with_size cons_r as before) ->
+      if cons_r.size_modifier != Size.used_once then
+        Size.update_once r.size cons_r.size_modifier;
+      if r.equal cons_r.key key then false
+      else if assoc_node r.equal key cons_r.rest != Nil then false
+      else
+        let (Cons_with_size cons_r as after) : ('a, 'b, [ `Cons_with_size ]) tdt
+            =
+          Cons_with_size
+            {
+              size_modifier = Size.(new_once r.size incr);
+              key;
+              value;
+              rest =
+                Cons
+                  { key = cons_r.key; value = cons_r.value; rest = cons_r.rest };
+            }
+        in
+        if Atomic_array.unsafe_compare_and_set r.buckets i (B before) (B after)
+        then (
+          Size.update_once r.size cons_r.size_modifier;
+          cons_r.size_modifier <- Size.used_once;
+          adjust_size t r mask true)
+        else try_add t key value (Backoff.once backoff)
+  | B (Resize _) ->
+      let _ = finish t (Atomic.get t) in
+      try_add t key value Backoff.default
+
+let[@inline] try_add t key value = try_add t key value Backoff.default
+
+(* *)
+
+let[@tail_mod_cons] rec dissoc t key = function
+  | Nil -> raise_notrace Not_found
+  | Cons r ->
+      if t key r.key then r.rest else Cons { r with rest = dissoc t key r.rest }
+
+let rec remove_node t key backoff =
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  match Atomic_array.unsafe_fenceless_get r.buckets i with
+  | B (Nil_with_size nil_r) ->
+      if nil_r.size_modifier != Size.used_once then (
+        Size.update_once r.size nil_r.size_modifier;
+        nil_r.size_modifier <- Size.used_once);
```

Comment:
Perhaps the clearing the `size_modifier` in this case might be best skipped.  The logic being that you clear it always after you introduce a new `size_modifier`.

Reply:
I actually thought the reason to clear the size_modifier value was so further  call of `update_once` on it will be avoided, but I guess it is actually to ... clear memory ?

Reply:
It is a bit of both.

Note that operations that create new `Nil_with_size { size_modifier }` objects where `size_modifier` isn't initially `Size.used_once` call `Size.update_once` and overwrite the `size_modifier` with `Size.used_once`.  So, there is usually only a tiny window of time during which the `size_modifier` is not `Size.used_once`.  Performing `r.size_modifier <- Size.used_once` is actually rather expensive, because it calls `caml_modify`, so, while it makes sense to do it once, to free a bit of space and avoid repeated calls to `Size.update_once`, it also makes sense to avoid doing that mutation multiple times.